private/bufpkg/bufconfig/buf_yaml_file.go

```diff
@@ -275,6 +300,20 @@ func (c *bufYAMLFile) ModuleConfigs() []ModuleConfig {
 	return slicesext.Copy(c.moduleConfigs)
 }
 
+func (c *bufYAMLFile) DefaultLintConfig() LintConfig {
+	if c.defaultLintConfig == nil {
+		return c.moduleConfigs[0].LintConfig()
+	}
+	return c.defaultLintConfig
+}
+
+func (c *bufYAMLFile) DefaultBreakingConfig() BreakingConfig {
+	if c.defaultBreakingConfig == nil {
+		return c.moduleConfigs[0].BreakingConfig()
```

Comment:
This isn't actually correct - this is just choosing a single configuration among potentially many and calling it the default, which is what we've tried to avoid pretty consistently with v2. I think what you're actually looking for in this PR is some concept of "if there is a single configuration, give it to me, otherwise tell me it is not present". What this logic looks like:

```func
SingleLintConfig() (LintConfig, bool)
```

- If v1, return the v1 configuration.
- If v2:
  - If there is one module with attached configuration, return the attached configuration. Else, return the top-level configuation.
  - If there is more than one module, return false

If you get false, do nothing, as we do right now.

Reply:
So if a user passes the following file to `--config`:

```
version: v2
lint:
  use:
    - WIRE
breaking:
  use:
    - FILE
modules:
  - path: foo
  - path: bar
    lint:
      use:
        - DEFAULT
      except:
        - IMPORT_USED
  - path: vendor
```

Wouldn't we want them to pick up the top-level configs for the default?

Reply:
We'd probably want to error, or at least print a warning along the lines of:

```
WARN Configuration {from flag,at $PATH was used} but this configuration had multiple modules. Buf cannot deduce which {lint,breaking} configuration to use with your image. Ignoring your configuration and using the default configuration instead.
```

I could see an argument to also modify the logic to "if >1 modules, return the top-level module", but that may just leave us open to even more support stuff in the long-term.

Going a different direction, we could take the position that for image inputs only, we always take the top-level configuration, and never look into specific module configs. This would mean that if you had the following:

```
version: v2
modules:
  - path: .
     lint: ...
```

The `modules[0].LintConfig` would not be picked up and propagated. The position here would be that module-specific configuration never applies to an image, and we treat configuration for an image as if it were a "new" module in the `modules` list.

This may be what `moduleConfigs[0]` does in practice for v1, but not for v2, but I'd also be comfortable with adopting this position, it seems intellectually consistent.

Reply:
I am a fan of the latter approach for images. For `v1`, the behaviour will stay the same since there should only ever be a single lint/breaking config. For `v2`,  we check for a top-level config, and if nothing is there, then we would use whatever the version defaults are. And yes, in the case you provided, where there is a single module, `path: .`, that would not be considered the default configuration. Changes incoming.